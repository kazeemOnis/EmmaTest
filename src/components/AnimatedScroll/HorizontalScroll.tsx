/* eslint-disable react-native/no-inline-styles */
import React from 'react';
import {FlatList} from 'react-native';
import {useNavigation} from '@react-navigation/native';
import {StackNavigationProp} from '@react-navigation/stack';
import styled from '../../theme';
import {Avatar} from '../../components';
import {OnboardingStackParamList} from '../../navigation/OnboardingNavigator/OnboardingNavigator';

interface HorizontalScrollProps {
  data: any[];
  dataType: 'team' | 'investor' | 'bank';
  activeIndex: number;
  imageWidth?: number;
  activeScroll: 'vertical' | 'horizontal';
  setActiveIndex: React.Dispatch<React.SetStateAction<number>>;
  setActiveScroll: React.Dispatch<
    React.SetStateAction<'vertical' | 'horizontal'>
  >;
}

const Separator = styled.View`
  width: 16px;
`;

export const HorizontalScroll = ({
  data,
  activeIndex,
  imageWidth,
  activeScroll,
  setActiveIndex,
  setActiveScroll,
  dataType,
}: HorizontalScrollProps) => {
  const navigation =
    useNavigation<StackNavigationProp<OnboardingStackParamList>>();
  const ref = React.useRef<FlatList>(null);
  // Default width is 75px, maximum is 150px, minimum width is 50px;
  const width =
    imageWidth && imageWidth <= 150 && imageWidth >= 50 ? imageWidth : 75;

  React.useEffect(() => {
    if (activeScroll === 'vertical') {
      ref.current?.scrollToIndex({
        index: activeIndex,
        animated: true,
        viewPosition: 0.5,
      });
    }
  }, [activeScroll, activeIndex]);

  return (
    <FlatList
      ref={ref}
      initialScrollIndex={activeIndex}
      data={data}
      horizontal
      showsHorizontalScrollIndicator={false}
      contentContainerStyle={{paddingHorizontal: 120}}
      onResponderMove={_event => {
        setActiveScroll('horizontal');
      }}
      onScroll={event => {
        const index = Math.abs(
          Math.round(event.nativeEvent.contentOffset.x / (width + 16)),
        );
        if (
          index >= 0 &&
          index < data.length &&
          activeScroll === 'horizontal'
        ) {
          setActiveIndex(index);
        }
      }}
      renderItem={({item, index}: {item: any; index: number}) => (
        <Avatar
          image={item.image}
          imageWidth={imageWidth}
          active={activeIndex === index}
          onPress={() =>
            navigation.navigate('Details', {id: item.id, type: dataType})
          }
        />
      )}
      ItemSeparatorComponent={Separator}
    />
  );
};
