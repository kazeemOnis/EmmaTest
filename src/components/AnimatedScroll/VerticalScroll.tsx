import React from 'react';
import {Dimensions, FlatList} from 'react-native';
import {Content} from '../../components';
import {Data} from '../../types';

interface AnimatedScrollProps {
  data: Data[];
  activeIndex: number;
  activeScroll: 'vertical' | 'horizontal';
  setActiveIndex: React.Dispatch<React.SetStateAction<number>>;
  setActiveScroll: React.Dispatch<
    React.SetStateAction<'vertical' | 'horizontal'>
  >;
}

export const VerticalScroll = ({
  data,
  activeIndex,
  activeScroll,
  setActiveIndex,
  setActiveScroll,
}: AnimatedScrollProps) => {
  const ref = React.useRef<FlatList>(null);

  React.useEffect(() => {
    if (activeScroll === 'horizontal') {
      ref.current?.scrollToIndex({
        index: activeIndex,
        animated: true,
      });
    }
  }, [activeScroll, activeIndex]);

  return (
    <FlatList
      ref={ref}
      data={data}
      initialScrollIndex={activeIndex}
      showsVerticalScrollIndicator={false}
      snapToAlignment="start"
      decelerationRate="fast"
      snapToInterval={Dimensions.get('window').height}
      onResponderMove={_event => {
        setActiveScroll('vertical');
      }}
      onScroll={event => {
        const index = Math.abs(
          Math.round(
            event.nativeEvent.contentOffset.y / Dimensions.get('window').height,
          ),
        );
        if (activeScroll === 'vertical') {
          setActiveIndex(index);
        }
      }}
      renderItem={({item}: {item: Data}) => (
        <Content
          title={item.title}
          subtext={item.subtext}
          content={item.content}
        />
      )}
    />
  );
};
