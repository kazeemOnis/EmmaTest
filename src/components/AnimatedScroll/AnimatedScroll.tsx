import React from 'react';
import styled from '../../theme';
import {HorizontalScroll} from './HorizontalScroll';
import {VerticalScroll} from './VerticalScroll';

interface AnimatedScrollProps {
  data: any[];
  dataType: 'team' | 'investor' | 'bank';
  imageWidth?: number;
}

const Wrapper = styled.View`
  padding-vertical: 24px;
`;

export const AnimatedScroll = ({
  data,
  imageWidth,
  dataType,
}: AnimatedScrollProps) => {
  const [index, setActiveIndex] = React.useState(0);
  const [activeScroll, setActiveScroll] = React.useState<
    'vertical' | 'horizontal'
  >('horizontal');

  return (
    <React.Fragment>
      <Wrapper>
        <HorizontalScroll
          data={data}
          dataType={dataType}
          activeIndex={index}
          activeScroll={activeScroll}
          setActiveIndex={setActiveIndex}
          setActiveScroll={setActiveScroll}
          imageWidth={imageWidth}
        />
      </Wrapper>
      <VerticalScroll
        data={data}
        activeIndex={index}
        activeScroll={activeScroll}
        setActiveIndex={setActiveIndex}
        setActiveScroll={setActiveScroll}
      />
    </React.Fragment>
  );
};
