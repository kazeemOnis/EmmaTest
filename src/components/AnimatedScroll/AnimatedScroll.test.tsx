import React from 'react';
import {cleanup, render} from '@testing-library/react-native';
import {createStackNavigator} from '@react-navigation/stack';
import {NavigationContainer} from '@react-navigation/native';
import {act} from 'react-test-renderer';
import {AnimatedScroll} from './AnimatedScroll';

afterEach(() => {
  cleanup();
});

const mockData = [
  {
    id: 1,
    title: 'Kazeem',
    subtext: 'React Native Engineer',
    image: 'https://images.unsplash.com/photo-1579353977828-2a4eab540b9a',
    content:
      'Experienced React Native Engineer with over 4 years of experience in building and shipping production ready apps to users',
  },
];

const TestComponent = () => {
  const TestStack = createStackNavigator();
  return (
    <NavigationContainer>
      <TestStack.Navigator>
        <TestStack.Screen
          name="Team"
          children={() => <AnimatedScroll data={mockData} dataType="team" />}
        />
      </TestStack.Navigator>
    </NavigationContainer>
  );
};

const setup = (overrides = {}) => {
  const props = {...mockData, ...overrides};
  const utils = render(<TestComponent />);
  return {...utils, props};
};

describe('Animated Scroll', () => {
  it('Should render Animated scroll', async () => {
    const {toJSON} = setup();
    await act(async () => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
