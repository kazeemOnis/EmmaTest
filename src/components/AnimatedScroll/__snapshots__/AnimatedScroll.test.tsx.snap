// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animated Scroll Should render Animated scroll 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "rgb(255, 255, 255)",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  Team
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          onGestureCancel={[Function]}
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          [
                            {
                              "paddingVertical": 24,
                            },
                          ]
                        }
                      >
                        <RCTScrollView
                          ItemSeparatorComponent={
                            {
                              "$$typeof": Symbol(react.forward_ref),
                              "attrs": [],
                              "inlineStyle": InlineStyle {
                                "rules": [
                                  "
  width: 16px;
",
                                ],
                              },
                              "render": [Function],
                              "shouldForwardProp": undefined,
                              "styledComponentId": "StyledNativeComponent",
                              "target": [Function],
                              "withComponent": [Function],
                            }
                          }
                          contentContainerStyle={
                            {
                              "paddingHorizontal": 120,
                            }
                          }
                          data={
                            [
                              {
                                "content": "Experienced React Native Engineer with over 4 years of experience in building and shipping production ready apps to users",
                                "id": 1,
                                "image": "https://images.unsplash.com/photo-1579353977828-2a4eab540b9a",
                                "subtext": "React Native Engineer",
                                "title": "Kazeem",
                              },
                            ]
                          }
                          getItem={[Function]}
                          getItemCount={[Function]}
                          horizontal={true}
                          initialScrollIndex={0}
                          keyExtractor={[Function]}
                          onContentSizeChange={[Function]}
                          onLayout={[Function]}
                          onMomentumScrollBegin={[Function]}
                          onMomentumScrollEnd={[Function]}
                          onResponderMove={[Function]}
                          onScroll={[Function]}
                          onScrollBeginDrag={[Function]}
                          onScrollEndDrag={[Function]}
                          removeClippedSubviews={false}
                          renderItem={[Function]}
                          scrollEventThrottle={50}
                          showsHorizontalScrollIndicator={false}
                          stickyHeaderIndices={[]}
                          viewabilityConfigCallbackPairs={[]}
                        >
                          <View>
                            <View
                              onFocusCapture={[Function]}
                              onLayout={[Function]}
                              style={
                                [
                                  {
                                    "flexDirection": "row",
                                  },
                                  null,
                                ]
                              }
                            >
                              <View
                                accessibilityState={
                                  {
                                    "busy": undefined,
                                    "checked": undefined,
                                    "disabled": undefined,
                                    "expanded": undefined,
                                    "selected": undefined,
                                  }
                                }
                                accessibilityValue={
                                  {
                                    "max": undefined,
                                    "min": undefined,
                                    "now": undefined,
                                    "text": undefined,
                                  }
                                }
                                accessible={true}
                                collapsable={false}
                                focusable={true}
                                onClick={[Function]}
                                onResponderGrant={[Function]}
                                onResponderMove={[Function]}
                                onResponderRelease={[Function]}
                                onResponderTerminate={[Function]}
                                onResponderTerminationRequest={[Function]}
                                onStartShouldSetResponder={[Function]}
                                style={
                                  {
                                    "alignItems": "center",
                                    "borderColor": "#5CD3CD",
                                    "borderRadius": 40,
                                    "borderWidth": 5,
                                    "height": 80,
                                    "justifyContent": "center",
                                    "opacity": 1,
                                    "width": 80,
                                  }
                                }
                                testID="avatar"
                              >
                                <View
                                  style={
                                    [
                                      {
                                        "borderRadius": 37.5,
                                        "height": 75,
                                        "width": 75,
                                      },
                                    ]
                                  }
                                  width={75}
                                >
                                  <Image
                                    radius={37.5}
                                    source={
                                      {
                                        "uri": "https://images.unsplash.com/photo-1579353977828-2a4eab540b9a",
                                      }
                                    }
                                    style={
                                      [
                                        {
                                          "borderRadius": 37.5,
                                          "height": "100%",
                                          "width": "100%",
                                        },
                                      ]
                                    }
                                  />
                                </View>
                              </View>
                            </View>
                          </View>
                        </RCTScrollView>
                      </View>
                      <RCTScrollView
                        data={
                          [
                            {
                              "content": "Experienced React Native Engineer with over 4 years of experience in building and shipping production ready apps to users",
                              "id": 1,
                              "image": "https://images.unsplash.com/photo-1579353977828-2a4eab540b9a",
                              "subtext": "React Native Engineer",
                              "title": "Kazeem",
                            },
                          ]
                        }
                        decelerationRate="fast"
                        getItem={[Function]}
                        getItemCount={[Function]}
                        initialScrollIndex={0}
                        keyExtractor={[Function]}
                        onContentSizeChange={[Function]}
                        onLayout={[Function]}
                        onMomentumScrollBegin={[Function]}
                        onMomentumScrollEnd={[Function]}
                        onResponderMove={[Function]}
                        onScroll={[Function]}
                        onScrollBeginDrag={[Function]}
                        onScrollEndDrag={[Function]}
                        removeClippedSubviews={false}
                        renderItem={[Function]}
                        scrollEventThrottle={50}
                        showsVerticalScrollIndicator={false}
                        snapToAlignment="start"
                        snapToInterval={1334}
                        stickyHeaderIndices={[]}
                        viewabilityConfigCallbackPairs={[]}
                      >
                        <View>
                          <View
                            onFocusCapture={[Function]}
                            onLayout={[Function]}
                            style={null}
                          >
                            <View
                              style={
                                [
                                  {
                                    "flexBasis": 0,
                                    "flexGrow": 1,
                                    "flexShrink": 1,
                                    "height": 1334,
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  [
                                    {
                                      "color": "white",
                                      "fontFamily": "Nunito",
                                      "fontSize": 24,
                                      "marginBottom": 4,
                                      "textAlign": "center",
                                    },
                                  ]
                                }
                                testID="contentTitle"
                              >
                                Kazeem
                              </Text>
                              <Text
                                style={
                                  [
                                    {
                                      "color": "white",
                                      "fontFamily": "Nunito",
                                      "fontSize": 16,
                                      "textAlign": "center",
                                    },
                                  ]
                                }
                                testID="contentSubtext"
                              >
                                React Native Engineer
                              </Text>
                              <Text
                                style={
                                  [
                                    {
                                      "color": "white",
                                      "fontFamily": "Nunito",
                                      "fontSize": 14,
                                      "marginTop": 16,
                                    },
                                  ]
                                }
                                testID="contentContent"
                              >
                                Experienced React Native Engineer with over 4 years of experience in building and shipping production ready apps to users
                              </Text>
                            </View>
                          </View>
                        </View>
                      </RCTScrollView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
